pilot/pkg/security/authn/v1beta1/policy_applier.go

```diff
@@ -343,111 +344,71 @@ func getMutualTLSMode(mtls *v1beta1.PeerAuthentication_MutualTLS) model.MutualTL
 func composePeerAuthentication(rootNamespace string, configs []*model.Config) *v1beta1.PeerAuthentication {
-	var meshPolicy, namespacePolicy *v1beta1.PeerAuthentication
+	var meshPolicy, namespacePolicy, workloadPolicy *v1beta1.PeerAuthentication
+	// Creation time associate with the selected workloadPolicy above. Intiial to max time (not set)
+	workloadPolicyCreationTime := time.Unix(1<<63-1, 0)
 
-	// Track number of workload level policies for early return.
-	numWorkloadLevelPolicies := 0
-
-	// First round to find mesh and namespace level policy.
 	for _, cfg := range configs {
 		spec := cfg.Spec.(*v1beta1.PeerAuthentication)
 		if cfg.Namespace == rootNamespace && spec.Selector == nil {
 			meshPolicy = spec
 		} else if spec.Selector == nil {
 			namespacePolicy = spec
 		} else if cfg.Namespace != rootNamespace {
-			numWorkloadLevelPolicies++
+			// Assign to the (selected) workloadPolicy, if it is not set or have a newer timestamp.
+			if workloadPolicy == nil || cfg.CreationTimestamp.Before(workloadPolicyCreationTime) {
+				workloadPolicy = spec
+				workloadPolicyCreationTime = cfg.CreationTimestamp
+			}
 		}
 	}
 
-	if meshPolicy == nil && namespacePolicy == nil && numWorkloadLevelPolicies == 0 {
+	if meshPolicy == nil && namespacePolicy == nil && workloadPolicy == nil {
+		// Return nil so that caller can fallback to apply alpha policy. Once we deprecate alpha API,
+		// this special case can be removed.
 		return nil
 	}
 
-	// Initial parentPolicy is set to a PERMISSIVE.
-	parentPolicy := v1beta1.PeerAuthentication{
+	// Initial outputPolicy is set to a PERMISSIVE.
+	outputPolicy := v1beta1.PeerAuthentication{
 		Mtls: &v1beta1.PeerAuthentication_MutualTLS{
 			Mode: v1beta1.PeerAuthentication_MutualTLS_PERMISSIVE,
 		},
 	}
 
-	if meshPolicy != nil && meshPolicy.Mtls != nil && meshPolicy.Mtls.Mode != v1beta1.PeerAuthentication_MutualTLS_UNSET {
-		// If mesh policy is define, update parentPolicy to mesh policy.
-		parentPolicy.Mtls = meshPolicy.Mtls
-	}
+	// Process in mesh, namespace, workload order to resolve inheritance (UNSET)
 
-	if namespacePolicy != nil && namespacePolicy.Mtls != nil && namespacePolicy.Mtls.Mode != v1beta1.PeerAuthentication_MutualTLS_UNSET {
-		// If namespace policy is define, update parent policy to namespace policy. This means namespace
-		// policy can completely overwrite mesh policy.
-		parentPolicy.Mtls = namespacePolicy.Mtls
+	if meshPolicy != nil && !isMtlsModeUnset(meshPolicy.Mtls) {
+		// If mesh policy is defined, update parentPolicy to mesh policy.
+		outputPolicy.Mtls = meshPolicy.Mtls
 	}
 
-	if numWorkloadLevelPolicies == 0 {
-		// There is no workload-level policy, return the parent policy.
-		return &parentPolicy
+	if namespacePolicy != nil && !isMtlsModeUnset(namespacePolicy.Mtls) {
+		// If namespace policy is defined, update output policy to namespace policy. This means namespace
+		// policy overwrite mesh policy.
+		outputPolicy.Mtls = namespacePolicy.Mtls
```

Comment:
wait, are we abandoning inherit stronger from namespace/mesh policy as well? i thought we only remove inherit for workload level policy.

Reply:
No, it's never before. There is at most 1 namespace level, so no conflict. Same as mesh. Narrower scope always win.